(PUP-9719) Fix Administrators Group Windows Permissions

gimmyxd opened the pull request.

gimmyxd commented:
jenkins please test this on windows10ent-64,windows10ent-32,windows2008r2-64,windows2012r2-64,windows2016-64,windows2019-64,windows2008-64,windows2012-64,windows2012r2-64,windows2012r2_wmf5-64,windows2012r2_core-64,windows7-64,windows81-64,windows10ent-32,windows10ent-32,windows2016-64,windows2016_core-64,windows2019_core-64,windows2019_ja-64a

puppetcla commented:
CLA signed by all contributors.

gimmyxd commented:
jenkins please test this on windows10ent-64,windows10ent-32,windows2008r2-64,windows2012r2-64,windows2016-64,windows2019-64,windows2008-64,windows2012-64,windows2012r2-64,windows2012r2_wmf5-64,windows2012r2_core-64,windows7-64,windows81-64,windows10ent-32,windows10ent-32,windows2016-64,windows2016_core-64,windows2019_core-64,windows2019_ja-64a

gimmyxd commented:
Tested with puppetlabs-acl, puppetlabs-registry, puppetlabs-iis, puppetlabs-reboot

gimmyxd commented on spec/unit/x509/cert_provider_spec.rb:
```diff
@@ -19,10 +19,11 @@ def expects_public_file(path)
+        an_object_having_attributes(sid: Puppet::Util::Windows::SID::BuiltinAdministrators, mask: 0x120089),
+        an_object_having_attributes(sid: current_sid, mask: 0x12019f),
+        an_object_having_attributes(sid: Puppet::Util::Windows::SID::BuiltinUsers, mask: 0x120089),
+        an_object_having_attributes(sid: Puppet::Util::Windows::SID::Everyone, mask: 0x120089)
```
Not sure if this set of permissions are right. addition to the ones before: `sid: Puppet::Util::Windows::SID::Everyone`

gimmyxd commented:
Need to limit the changes the and minimize the impact on existing functionality

gimmyxd commented:
jenkins please test this on windows10ent-64,windows10ent-32,windows2008r2-64,windows2012r2-64,windows2016-64,windows2019-64,windows2008-64,windows2012-64,windows2012r2-64,windows2012r2_wmf5-64,windows2012r2_core-64,windows7-64,windows81-64,windows10ent-32,windows10ent-32,windows2016-64,windows2016_core-64,windows2019_core-64,windows2019_ja-64a

gimmyxd commented on lib/puppet/file_system/windows.rb:
```diff
@@ -177,14 +200,23 @@ def set_dacl(path, dacl)
     ].uniq.map do |sid|
       dacl.allow(sid, FULL_CONTROL)
+        permission = case sid
+                     when Puppet::Util::Windows::SID::LocalSystem
```
Should SYSTEM always have FULL_CONTROL instead of owner_permission?

ciprianbadescu commented on acceptance/tests/windows/PUP-9719_windows_system_first_pa_run.rb:
```diff
@@ -0,0 +1,56 @@
+test_name "PUP-9719 Windows First Agent run as SYSTEM sets cache file permissions correctly" do
+  tag 'risk:medium',
+    'audit:medium',
+    'audit:integration' # exec resource succeeds when the `exit_code` parameter
```
is this comment needed?

ciprianbadescu commented on lib/puppet/file_system/windows.rb:
```diff
@@ -4,6 +4,8 @@
   FULL_CONTROL = Puppet::Util::Windows::File::FILE_ALL_ACCESS
   FILE_READ = Puppet::Util::Windows::File::FILE_GENERIC_READ
+  FILE_WRITE = Puppet::Util::Windows::File::FILE_GENERIC_WRITE
+  FILE_RW = (FILE_READ | FILE_WRITE)
```
cosmetic, it could be FILE_READ_WRITE to be consistent with FILE_READ and FILE_WRITE

ciprianbadescu commented on lib/puppet/file_system/windows.rb:
```diff
@@ -134,10 +137,30 @@ def replace_file(path, mode = nil)
              secure_dacl(current_sid)
+           when 0660
+             dacl = secure_dacl(current_sid, FILE_RW, FILE_RW)
+             dacl
```
is `secure_dacl(current_sid, FILE_RW, FILE_RW)` enough?

gimmyxd replied:
660 is read and write for user and group, so this should be enough

ciprianbadescu commented on lib/puppet/file_system/windows.rb:
```diff
@@ -134,10 +137,30 @@ def replace_file(path, mode = nil)
              dacl
            when 0640, 0600
              secure_dacl(current_sid)
```
This will default to secure_dacl(current_sid, FULL_CONTROL, FULL_CONTROL), is this to permissive?

gimmyxd replied:
It is but i don't want to change the existing functionality, just add the new modes. I'll fill a new ticket to proper solve this.

joshcooper replied:
`current_sid` is already granted `FULL_CONTROL` in https://github.com/puppetlabs/puppet/pull/7807/commits/a43cdf5ac35ae61bf38889052276e8971229a4b7#diff-d56421f468d3cfe2988bff9fb021f192R210

Also, the current owner of a file has the `WRITE_DAC` permission, so they can always change the DACL and grant themselves `FULL_CONTROL`.

ciprianbadescu commented on lib/puppet/file_system/windows.rb:
```diff
@@ -177,14 +200,23 @@ def set_dacl(path, dacl)
+                       group_permission
+                     else
+                       owner_permission
+                     end
```
could this be written as:
{{{
when Puppet::Util::Windows::SID::BuiltinAdministrators
  group_permission
else
 owner_permission
  }}}

gimmyxd commented:
jenkins please test this on windows10ent-64,windows10ent-32,windows2008r2-64,windows2012r2-64,windows2016-64,windows2019-64,windows2008-64,windows2012-64,windows2012r2-64,windows2012r2_wmf5-64,windows2012r2_core-64,windows7-64,windows81-64,windows10ent-32,windows10ent-32,windows2016-64,windows2016_core-64,windows2019_core-64,windows2019_ja-64a

Dorin-Pleava commented:
Tested version 6.4.5 on windows 2016 without the fix and the access denied error occurred when installing via scheduled tasks, tested the fix and the error no longer occurred.

joshcooper commented:
Kicking Travis

joshcooper commented on acceptance/tests/windows/PUP-9719_windows_system_first_pa_run.rb:
```diff
@@ -0,0 +1,53 @@
+    end
+
+    step "Create and run a scheduled task on System Account." do
+      on agent, "schtasks /create /tn PuppetSystemRun /RL HIGHEST /RU SYSTEM /F /SC ONCE /ST 23:59 /TR 'cmd /c \"%ProgramFiles%\\Puppet Labs\\Puppet\\bin\\puppet.bat\" agent -t >> c:\\Windows\\Temp\\Puppet-System-Run.log 2>&1'"
```
How does setting start time to "23:59" ensure the agent runs "now"? Based on https://docs.microsoft.com/en-us/windows-server/administration/windows-commands/schtasks#BKMK_now, it sounds like we'd have to get the current time, and add a few seconds, but I think the time needs to be in the future, as if we "miss" the time, then it won't run until tomorrow?

Can we use `psexec` to run the agent now, instead of dealing with scheduled tasks?

gimmyxd replied:
This just create the task, running the task now is done on the next line:
```on agent, "schtasks /run /tn PuppetSystemRun"```

joshcooper replied:
Are there timing issues where the task scheduler might run the task (if we're close to 23:59) and then we try to run the task again?

joshcooper commented on acceptance/tests/windows/PUP-9719_windows_system_first_pa_run.rb:
```diff
@@ -0,0 +1,53 @@
+    end
+
+    step "Wait for Puppet Agent run to complete" do
+      last_puppet_run  = File.join(statedir, "/last_run_summary.yaml")
```
nit `File.join` adds slashes between path components, so it isn't needed here:

joshcooper commented on lib/puppet/file_system/windows.rb:
```diff
@@ -177,14 +200,17 @@ def set_dacl(path, dacl)
     ].uniq.map do |sid|
       dacl.allow(sid, FULL_CONTROL)
+      permission = owner_permission
+      permission = group_permission if Puppet::Util::Windows::SID::BuiltinAdministrators
```
Is this supposed to be conditional? I think right now `BuiltinAdministrators` is always defined, so `permission` is overwritten with `group_permission`.

joshcooper commented on lib/puppet/file_system/windows.rb:
```diff
@@ -177,14 +200,17 @@ def set_dacl(path, dacl)
       dacl.allow(sid, FULL_CONTROL)
+      permission = owner_permission
+      permission = group_permission if Puppet::Util::Windows::SID::BuiltinAdministrators
+      dacl.allow(sid, permission)
```
I'm not sure I follow this. If `sid` is granted `FULL_CONTROL` above on line 210, then this line doesn't grant any additional permissions.

gimmyxd replied:
Removed `dacl.allow(sid, FULL_CONTROL)`

joshcooper replied:
`LocalSystem` and `BuiltinAdministrators` should always have `FULL_CONTROL`, so this method shouldn't change.

gimmyxd replied:
This means that `current_sid` will always have full control. Looking at `0440` is it worth changing `current_sid` permissions?

joshcooper commented on spec/unit/transaction/persistence_spec.rb:
```diff
@@ -146,16 +146,7 @@ def write_state_file(contents)
-      else
-        expect { persistence.save }.to raise_error(Errno::EISDIR, /Is a directory/)
-      end
+      expect { persistence.save }.to raise_error(Errno::EISDIR, /Is a directory/)
```
Curious why this changed?

joshcooper replied:
ah cool, makes sense

gimmyxd commented on acceptance/tests/windows/PUP-9719_windows_system_first_pa_run.rb:
```diff
@@ -0,0 +1,33 @@
+    end
+
+    step "Run puppet as System Account" do
+      on agent, 'psexec -s cmd /c "C:\\Program Files\\Puppet Labs\\Puppet\\bin\\puppet.bat" agent -t >> C:\\Windows\\Temp\\Puppet-System-Run.log 2>&1'
```
This is still not working as expected, the test pass even without the permission changes.

gimmyxd commented:
@joshcooper i updated the code to address your remarks. I still need to update the acceptance test, but you can have a look on the updated functionality.

gimmyxd commented:
I'll also address the travis commit message failure tomorrow

joshcooper commented on lib/puppet/file_system/windows.rb:
```diff
@@ -134,10 +137,30 @@ def replace_file(path, mode = nil)
              secure_dacl(current_sid)
+           when 0660
+             dacl = secure_dacl(current_sid, FILE_READ_WRITE, FILE_READ_WRITE)
+             dacl
```
I would treat `0660` the same as `0640` above.

joshcooper commented on lib/puppet/file_system/windows.rb:
```diff
@@ -134,10 +137,30 @@ def replace_file(path, mode = nil)
+             dacl = secure_dacl(current_sid, FILE_READ_WRITE, FILE_READ_WRITE)
+             dacl.allow(Puppet::Util::Windows::SID::BuiltinUsers, FILE_READ)
+             dacl.allow(Puppet::Util::Windows::SID::Everyone, FILE_READ)
+             dacl
```
I would drop `Everyone` and only set `BuiltUsers` to read or readwrite, as that's what Windows does when creating a file generally. Same comment below.

joshcooper commented on lib/puppet/file_system/windows.rb:
```diff
@@ -134,10 +137,30 @@ def replace_file(path, mode = nil)
+             dacl = secure_dacl(current_sid, FILE_READ, FILE_READ)
+             dacl.allow(Puppet::Util::Windows::SID::BuiltinUsers, FILE_READ)
+             dacl.allow(Puppet::Util::Windows::SID::Everyone, FILE_READ)
+             dacl
```
I would drop `0444` since it's never called from anywhere besides specs.

gimmyxd commented:
@joshcooper 
- I changed the scheduled task start time on the next day, then force start it now to avoid race conditions.
- i reverted `secure_dacl` method to initial implementation and added only the new modes that used by puppet: 0660(which behaves like 0640 and 0600)
- also added 0440 which i'm not sure if is worth changing `current_sid` permission to only read. Right now is behaving like 0660, 0640 and 0600

gimmyxd commented:
jenkins please test this on windows10ent-64,windows10ent-32,windows2008r2-64,windows2012r2-64,windows2016-64,windows2019-64,windows2008-64,windows2012-64,windows2012r2-64,windows2012r2_wmf5-64,windows2012r2_core-64,windows7-64,windows81-64,windows10ent-32,windows10ent-32,windows2016-64,windows2016_core-64,windows2019_core-64,windows2019_ja-64a

gimmyxd commented on spec/unit/file_system_spec.rb:
```diff
@@ -1065,12 +1065,60 @@ def increment_counter_in_multiple_processes(file, num_procs, options)
+            an_object_having_attributes(sid: Puppet::Util::Windows::SID::BuiltinAdministrators, mask: 0x1f01ff),
+            an_object_having_attributes(sid: current_sid, mask:  0x1f01ff)
+          )
+        end
```
I could DRY this after Jenkins runs, using a share example or an array of modes.

gimmyxd commented:
This will need to wait for packaging fix to be able to run acceptance test.

joshcooper commented:
jenkins please test this on windows10ent-64,windows10ent-32,windows2008r2-64,windows2012r2-64,windows2016-64,windows2019-64,windows2008-64,windows2012-64,windows2012r2-64,windows2012r2_wmf5-64,windows2012r2_core-64,windows7-64,windows81-64,windows10ent-32,windows10ent-32,windows2016-64,windows2016_core-64,windows2019_core-64,windows2019_ja-64a
